### PR TITLE
Codechange: introduce a framework for all our timers

### PR DESCRIPTION
**Pull request by TrueBrain**
## Motivation / Problem

We have a few things we would like to address:
- Autosaves based on real-time, not in-game time
- Similar for LinkGraphs

Additionally, we would like to split up game-time in "calendar" and "economy", to allow speeding up one and not the other.
This is all a bit tricky in the current code-base, as we track these kind of things all over the place.

So, I was like: can't we make a single system for Interval and Timeout timers, where you can just say based on what "clock" you want it to work.

This PR is a result of that idea. It just puts down the foundation, and converts game-date and window-based timers.

## Description

`IntervalTimer` and `TimeoutTimer` use RAII, and can be used to replace all the time-based timeouts, lag-detection, "execute every N" we have.
As it uses RAII, you can safely use it as static variable, class member, temporary variable, etc. As soon as it goes out-of-scope, it will be safely removed.
This allows for much easier to read code when it comes to intervals.

This replaces `GUITimer` for Windows.

Additional, for all Window related function I replaced `MILLISECONDS_PER_TICK` to `30`, as that is actual the intent. The Window system is completely uncoupled from the game-ticks, so it shouldn't make the suggestion it is different.

This PR is just the start of things. It got a bit bigger than I would like, but I wanted to show that we can use this to replace all oddities we have in our code. As the Window system has the most oddities, I replaced all time-related events there with this new system. Sometimes it can be a bit hard to see if the work I did is correct, as some code is just plain weird. I did my best.
Additionally, to make sure we can convert our game-date related timers too, tackled those. Found some fun issues with them, and addressed those.

Future work would include:
- Replace all lag-related network detection in a similar way.
- Move autosave / linkgraph to real-time, instead of game-time.
And I am sure I miss many more.

### O(n)

Every timer registered is checked on every progression of time. (read: `O(n)`). This is strictly seen not needed, and we could use a priority-tree, to check if only the top entry progressed sufficiently.
But, to not overcomplicate a relative few entries on a list, I didn't go that route. This PR puts down the foundation; we can always extend with this at a later time, within the interface of the current implementation.

### Templating

This PR makes strong use of templating; in theory we could split every time up in their own function, and do away with the templating. I mostly went for this route, to make it easier to switch from one time to the other, and to make sure all timers look the same, no matter what their trigger is.

### Magic numbers (resolved)

Many places use a `std::chrono`, instead of an enum for their timings. This was already there, or implied, but in theory it should be put in an enum. This PR only was already long enough, that I decided not to do that in this PR.
EDIT: after talking this over, decided to remove all enum references instead. This makes more sense, as this keeps the information local. The documentation of the enums were .. non-descriptive at best. This way the context is clear instantly.

### Related bugs

When doing a PR like this, you can't help but finding all kind of weird bugs. Like the setting `errmsg_duration` is in 3-second intervals, not 1-second what the tooltip suggests. Framerate graphs jump all over the place.
To keep a focus, I pushed those all out of scope. This PR should be as broken on these subjects as it was before this PR.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/company_gui.cpp [hunk 1]
```diff
@@ -2674,11 +2679,10 @@ struct CompanyWindow : Window
 		}
 	}
 
-	void OnHundredthTick() override
-	{
-		/* redraw the window every now and then */
+	/** Redraw the window on a regular interval. */
+	IntervalTimer<TimerWindow> redraw_interval = {std::chrono::seconds(3), [this](auto) {
```
Comment (LordAro) [inline src/company_gui.cpp:2688]: Does the actual time value ever vary? Is it likely to? Perhaps it could be a template parameter instead?

Comment (PeterN) [reply src/company_gui.cpp:2688]: There is at least one case where the timer is scaled by scaled font size.

**Comment by PeterN** [comment]
Regarding magic numbers and enums, as it's now possible to put the time, in ms, directly inside the timer declaration, I think the old way of declaring a constant or enum for the timer, somewhere else, is less useful.

File: src/company_gui.cpp (hunk 1 above)
Comment (TrueBrain) [reply src/company_gui.cpp:2688]: There is at least one instance of `std::chrono::seconds(3 * _settings_client.gui.errmsg_duration)`, and I suspect there will be a few more over time.

So yeah, sadly, no template parameter :(

**Comment by PeterN** [comment]
Is there a way to ensure timers will run in a specific order (in case of dependencies) or is the answer to that just to use one timer?

**Comment by PeterN** [comment]
I think it would only matter for game-loop stuff, for general desync-safety.

**Comment by TrueBrain** [comment]
> I think it would only matter for game-loop stuff, for general desync-safety.

This triggered me to make a few changes. The timers are now stored in a `std::set`, and sorted. This means it should be much more deterministic what is happening in the game, both for Window as for the Game stuff.

File: src/engine.cpp [hunk 2]
```diff
@@ -1110,6 +1112,11 @@ void EnginesMonthlyLoop()
 	}
 }
 
+static IntervalTimer<TimerGameCalendar> _engines_monthly({TimerGameCalendar::MONTH, TimerGameCalendar::Priority::ENGINE}, [](auto)
```
Comment (rubidium42) [inline src/engine.cpp:1115]: Shouldn't this line practically replace the `EnginesMonthlyLoop()` like it does for `EnginesDailyLoop()`?

Comment (TrueBrain) [reply src/engine.cpp:1115]: Yes, except for some reason this function is also called from the cheats menu.

PS: similar with `ResetSignalVariant` btw.

File: src/timer/timer_game_calendar.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * The amount of days in a month depends on the month and year (leap-years).
+ * There are always 74 ticks in a day (and with 27ms, this makes 1 day 1.998 seconds).
+ *
+ * IntervalTimer and TimeoutTimer based on this Timer are a bit unusual, as their count is always one.
+ * You create those timers based on a transition: a new tick, a new day, a new month or a new year.
+ *
+ * Additionally, you need to set a priority. To ensure deterministic behaviour, events are executed
+ * in priority. It is important that if you assign NONE, you do not use Random() in your callback.
+ * Other than that, make sure you only set one callback per priority.
```
Comment (frosch123) [inline src/timer/timer_game_calendar.h:24]: This is not specific to Random().
All non-GUI-code handlers should be sorted, so all priorities should have a single handler only.
Currently Priority::ENGINE is used twice, and I can imagine that already desyncs. Engine and track-types become available depending on each other, ...

Comment (TrueBrain) [reply src/timer/timer_game_calendar.h:24]: They aren't used twice. Well, strictly seen, but one time on Day, and one time on Month. That order is well defined. So it is deterministic currently, and I wouldn't see where any desync would be possible currently :)

Did you overlook that they are on different triggers, or am I missing something?

Comment (frosch123) [reply src/timer/timer_game_calendar.h:24]: Ah yes, I missed that Day/Month thing. Can there be an assertion in the gameloop timer, that the handlers with priority are strictly ordered?

Comment (TrueBrain) [reply src/timer/timer_game_calendar.h:24]: can do!

I also planned to do it, but forgot .. so tnx for pointing out it was missing :D

As if this fails means that a developer made a big boo-boo, the code only runs for assert-enabled builds. Means I can just `assert` to detect issues :)

**Comment by LordAro** [comment]
Tempted to just approve this.

Have you profiled at all, to confirm that all the fun global state doesn't slow things down at all?

**Comment by TrueBrain** [comment]
> Tempted to just approve this.
> 
> Have you profiled at all, to confirm that all the fun global state doesn't slow things down at all?

Ran several tests:

- Empty map, before and after this PR: no difference (both with and without fast-forward)
- Title-screen, before and after this PR: no difference (both with and without fast-forward)
- Clang thread-sanitizer: no reports

Although I didn't expect any performance issues (as we replace GUITimers with something similar and game-time related events are rare), I can also not measure any difference.

So I think we are all good here.

**Comment by PeterN** [comment]
Sorry, couldn't figure out how to copy & paste this as text... 

image

**Comment by TrueBrain** [comment]
> Sorry, couldn't figure out how to copy & paste this as text...
> 
> image
> 
> And one more.
> 
> image

Fixed!

**Comment by LordAro** [comment]
Did a little bit of profiling myself using `perf`

There's definitely some amount of performance impact...

Taking opntitle.dat (30000 ticks, repeated 5 times) & wentbourne.sav (3000 ticks, no repeats)

Linux, compiled in release mode - master @ 646a7e625b (buildrel), date-time-rework @ 1ede693e65 (buildrel2)

```
 > perf stat -r5 ./buildrel/openttd -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat 

 Performance counter stats for './buildrel/openttd -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat' (5 runs):

         29,563.91 msec task-clock:u                     #    4.990 CPUs utilized               ( +- 14.18% )
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
            69,685      page-faults:u                    #    3.940 K/sec                       ( +- 14.06% )
    98,115,471,425      cycles:u                         #    5.548 GHz                         ( +- 14.15% )
   179,000,427,381      instructions:u                   #    3.04  insn per cycle              ( +- 14.15% )
    26,804,484,363      branches:u                       #    1.516 G/sec                       ( +- 14.15% )
       721,462,981      branch-misses:u                  #    4.49% of all branches             ( +- 14.15% )

            5.9246 +- 0.0320 seconds time elapsed  ( +-  0.54% )

 > perf stat -r5 ./buildrel2/openttd -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat 

 Performance counter stats for './buildrel2/openttd -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat' (5 runs):

         34,586.48 msec task-clock:u                     #    4.988 CPUs utilized               ( +- 14.20% )
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
            69,750      page-faults:u                    #    3.380 K/sec                       ( +- 14.21% )
   116,449,386,463      cycles:u                         #    5.643 GHz                         ( +- 14.19% )
   219,830,916,148      instructions:u                   #    3.16  insn per cycle              ( +- 14.17% )
    36,147,753,437      branches:u                       #    1.752 G/sec                       ( +- 14.17% )
       801,839,343      branch-misses:u                  #    3.71% of all branches             ( +- 14.18% )

            6.9337 +- 0.0426 seconds time elapsed  ( +-  0.62% )

```

```
 > perf stat ./buildrel/openttd -x -snull -mnull -vnull:ticks=3000 -g ./save/globalsave/wentbourne.sav

 Performance counter stats for './buildrel/openttd -x -snull -mnull -vnull:ticks=3000 -g ./save/globalsave/wentbourne.sav':

        118,058.15 msec task-clock:u                     #    1.000 CPUs utilized             
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
            41,633      page-faults:u                    #  352.648 /sec                      
   442,205,883,147      cycles:u                         #    3.746 GHz                       
   562,435,201,628      instructions:u                   #    1.27  insn per cycle            
    98,967,160,301      branches:u                       #  838.292 M/sec                     
     2,258,783,701      branch-misses:u                  #    2.28% of all branches           

     118.082006900 seconds time elapsed

     117.115598000 seconds user
       0.484487000 seconds sys


 > perf stat ./buildrel2/openttd -x -snull -mnull -vnull:ticks=3000 -g ./save/globalsave/wentbourne.sav

 Performance counter stats for './buildrel2/openttd -x -snull -mnull -vnull:ticks=3000 -g ./save/globalsave/wentbourne.sav':

        152,760.85 msec task-clock:u                     #    1.000 CPUs utilized             
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
            41,570      page-faults:u                    #  272.125 /sec                      
   567,092,138,306      cycles:u                         #    3.712 GHz                       
   854,986,197,789      instructions:u                   #    1.51  insn per cycle            
   167,392,857,434      branches:u                       #    1.096 G/sec                     
     2,366,130,587      branch-misses:u                  #    1.41% of all branches           

     152.778481728 seconds time elapsed

     151.745360000 seconds user
       0.521426000 seconds sys

```

**Comment by TrueBrain** [comment]
I tried to reproduce your results, but I can't get even close to it.

First run is master (730687080ab2aa0bf62e9282bd9ac90c75da5f5d), second is my PR (1ede693e6515c94c445aaed51e9f8fa5f079e8c4).

Mostly, it is important to note I need about 8 times less instructions to execute the same amount of ticks. As we are talking on Discord/IRC, I really hope it is due to the `-X`, but this is scaring me :)

```
$ perf stat -r5 ./openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ../media/baseset/opntitle.dat

 Performance counter stats for './openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ../media/baseset/opntitle.dat' (5 runs):

          3,795.74 msec task-clock:u              #    1.153 CPUs utilized            ( +-  1.00% )
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
            35,016      page-faults:u             #    0.009 M/sec                    ( +-  0.65% )
    16,160,734,413      cycles:u                  #    4.258 GHz                      ( +-  0.89% )
    28,812,047,147      instructions:u            #    1.78  insn per cycle           ( +-  0.23% )
     5,251,025,809      branches:u                # 1383.401 M/sec                    ( +-  0.23% )
       125,279,800      branch-misses:u           #    2.39% of all branches          ( +-  0.40% )

            3.2915 +- 0.0296 seconds time elapsed  ( +-  0.90% )
```

vs

```
$ perf stat -r5 ./openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ../media/baseset/opntitle.dat

 Performance counter stats for './openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ../media/baseset/opntitle.dat' (5 runs):

          4,021.77 msec task-clock:u              #    1.148 CPUs utilized            ( +-  6.26% )
                 0      context-switches:u        #    0.000 K/sec                  
                 0      cpu-migrations:u          #    0.000 K/sec                  
            34,940      page-faults:u             #    0.009 M/sec                    ( +-  0.35% )
    16,832,544,841      cycles:u                  #    4.185 GHz                      ( +-  5.36% )
    28,976,702,152      instructions:u            #    1.72  insn per cycle           ( +-  1.70% )
     5,279,039,315      branches:u                # 1312.616 M/sec                    ( +-  1.64% )
       126,758,033      branch-misses:u           #    2.40% of all branches          ( +-  2.37% )

             3.502 +- 0.232 seconds time elapsed  ( +-  6.64% )
```

**Comment by LordAro** [comment]
Afraid not:

```
 > perf stat -r5 ./buildrel/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat 

 Performance counter stats for './buildrel/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat' (5 runs):

         17,603.60 msec task-clock:u                     #    5.904 CPUs utilized               ( +- 14.12% )
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
           171,775      page-faults:u                    #   16.254 K/sec                       ( +- 14.22% )
    64,320,086,533      cycles:u                         #    6.086 GHz                         ( +- 14.12% )
   114,448,978,420      instructions:u                   #    2.96  insn per cycle              ( +- 14.13% )
    19,704,155,280      branches:u                       #    1.865 G/sec                       ( +- 14.13% )
       591,615,013      branch-misses:u                  #    5.00% of all branches             ( +- 14.14% )

            2.9814 +- 0.0113 seconds time elapsed  ( +-  0.38% )

 > perf stat -r5 ./buildrel2/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat 

 Performance counter stats for './buildrel2/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat' (5 runs):

         20,493.62 msec task-clock:u                     #    5.761 CPUs utilized               ( +- 14.14% )
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
           174,002      page-faults:u                    #   14.147 K/sec                       ( +- 14.17% )
    75,350,379,955      cycles:u                         #    6.126 GHz                         ( +- 14.14% )
   140,140,451,927      instructions:u                   #    3.10  insn per cycle              ( +- 14.14% )
    25,966,873,534      branches:u                       #    2.111 G/sec                       ( +- 14.14% )
       626,430,131      branch-misses:u                  #    4.02% of all branches             ( +- 14.14% )

           3.55735 +- 0.00478 seconds time elapsed  ( +-  0.13% )
```

wentbourne similar (116,497ms vs 148,419ms)

**Comment by LordAro** [comment]
One reconfigure & recompile from scratch later... and it's not looking good for me

```
 > perf stat -r5 ./buildrel-master/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat 

 Performance counter stats for './buildrel-master/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat' (5 runs):

         20,495.62 msec task-clock:u                     #    5.758 CPUs utilized               ( +- 14.08% )
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
           175,113      page-faults:u                    #   14.164 K/sec                       ( +- 14.20% )
    75,330,485,498      cycles:u                         #    6.093 GHz                         ( +- 14.11% )
   141,290,665,639      instructions:u                   #    3.12  insn per cycle              ( +- 14.12% )
    26,181,598,341      branches:u                       #    2.118 G/sec                       ( +- 14.12% )
       627,130,708      branch-misses:u                  #    3.99% of all branches             ( +- 14.13% )

            3.5597 +- 0.0260 seconds time elapsed  ( +-  0.73% )

 > perf stat -r5 ./buildrel/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat 

 Performance counter stats for './buildrel/openttd -X -c empty/empty.cfg -x -snull -mnull -vnull:ticks=30000 -g ./media/baseset/opntitle.dat' (5 runs):

         20,480.86 msec task-clock:u                     #    5.759 CPUs utilized               ( +- 14.07% )
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
           174,221      page-faults:u                    #   14.150 K/sec                       ( +- 14.10% )
    75,428,243,777      cycles:u                         #    6.126 GHz                         ( +- 14.10% )
   140,115,508,926      instructions:u                   #    3.09  insn per cycle              ( +- 14.12% )
    25,962,820,371      branches:u                       #    2.109 G/sec                       ( +- 14.12% )
       626,215,023      branch-misses:u                  #    4.02% of all branches             ( +- 14.11% )

            3.5566 +- 0.0225 seconds time elapsed  ( +-  0.63% )
```

**Comment by TrueBrain** [comment]
A bit more fundamental answer to the question: how about the performance?

image

Basically, worst-case, this makes the game 0.23% slower. But as in there is also the time to handled the timers, it is significantly less than that. But as it used to be throughout the code, that number is hard to estimate. But it can never degrade more than 0.23% after this PR.

**Review by PeterN — APPROVED** [review]

**Pull request merged**